### PR TITLE
Enhance: metadata annotation for structs

### LINKED ISSUE #459: Allow rust struct generating freezed class in dart
In rust we often define:
```rust
struct UserId(value: u64);
```
And use UserId in HashMap as keys.

Since tuple is not well supported in dart, we can define:
```rust
struct UserId { value: u64 }
```
And use it in maps/dicts as keys

But frb currently doesn't generate `==` and `hashCode` method for UserId in dart/flutter, so we can't use UserId as keys in dart, we had to unwrap it and use u64 as key type, it's not quit convient.

If we support generate @freezed (or even better any annotation) on struct, then we got `==` and `hashCode`, and can use it directly in maps.

### PR DESCRIPTION
**Pull request by alanlzhang**
close #459

### CODE REVIEW

**Comment by alanlzhang** [comment]
CHANGELOG.md updated.

**Comment by alanlzhang** [comment]
@fzyzcjy Please approve the workflow since this is my first PR for frb ^_^

File: frb_codegen/src/ir/import.rs [hunk 1]
```diff
@@ -0,0 +1,5 @@
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+pub struct IrImport {
```
Comment (fzyzcjy) [inline frb_codegen/src/ir/import.rs:2]: After reading example, I realized it is not "import" in Rust, but instead in Dart. Shall we rename `IrImport -> IrDartImport` and `IrAnnotation -> IrDartAnnotation`?

File: frb_example/pure_dart/rust/src/api.rs [hunk 2]
```diff
@@ -537,3 +537,15 @@ pub fn get_complex_array() -> [Point; 2] {
 pub fn get_usize(u: usize) -> usize {
     u
 }
+
+/// Example for @freezed and @meta.immutable
+#[frb(metadata=("freezed", "immutable" import "package:meta/meta.dart" as meta))]
```
Comment (fzyzcjy) [inline frb_example/pure_dart/rust/src/api.rs:542]: I guess this means *annotations* for dart classes? If so, what about remaming "metadata=..." to "dart_annotation=..."

File: frb_example/pure_dart/dart/lib/main.dart [hunk 3]
```diff
@@ -372,6 +368,10 @@ void main(List<String> args) async {
     customized.nonFinalField = "changed";
     expect(customized.nonFinalField, "changed");
 
+    print('dart call next_user_id');
```
Comment (fzyzcjy) [inline frb_example/pure_dart/dart/lib/main.dart:371]:
```suggestion
    print('dart call next_user_id to test metadata annotations');
```

File: frb_codegen/src/parser/mod.rs [hunk 4]
```diff
@@ -213,6 +215,137 @@ fn extract_comments(attrs: &[Attribute]) -> Vec<IrComment> {
...
+            .collect();
+        Ok(Self(imports))
+    }
+}
+
+enum FrbOption {
+    Mirror(MirrorOption),
+    NonFinal,
+    Metadata(NamedOption<frb_keyword::metadata, MetadataAnnotations>),
+}
```
Comment (fzyzcjy) [inline frb_codegen/src/parser/mod.rs:316]: If I understand correctly, you successfully parsed *all* rust macros? That sounds great!

Seems that, the mirror/nonfinal/... already have their own parsers to parse them. Thus are we duplicating works here? Maybe we can refactor those?

**Comment by fzyzcjy** [comment]
Great work! Only some small changes and I am ready to merge it.

File: frb_codegen/src/parser/mod.rs (hunk 4 above)
Comment (alanlzhang) [reply frb_codegen/src/parser/mod.rs:316]: Yes, I try to make an uniform way to parse parameters in "[frb]" macros. I think we can refactor code to use it. But I am new here, so I think it's better to move things slowly and do it later. ^_^

File: frb_codegen/src/ir/import.rs (hunk 1 above)
Comment (alanlzhang) [reply frb_codegen/src/ir/import.rs:2]: Sure. :-)

**Comment by fzyzcjy** [comment]
Oh, one more thing: Would you mind add a subsection in the documentation :) For example, sibling to: http://cjycode.com/flutter_rust_bridge/feature/handler.html

File: frb_example/pure_dart/rust/src/api.rs (hunk 2 above)
Comment (fzyzcjy) [reply frb_example/pure_dart/rust/src/api.rs:542]: Good point! I did not realized they call the `int` in `int a = ...` a type annotation before. So `dart_metadata` is quite reasonable.

File: frb_codegen/src/parser/mod.rs (hunk 4 above)
Comment (fzyzcjy) [reply frb_codegen/src/parser/mod.rs:316]: Sure. Look forward to your future PRs!

**Comment by alanlzhang** [comment]
Ok. And I think just add some description in "struct" feature page because we can only add dart metadata annotations on rust struct.

**Comment by fzyzcjy** [comment]
Agree. Btw you can "resolve conversation" above when done.

**Comment by alanlzhang** [comment]
Add some line to the lang_struct.md.

**Comment by fzyzcjy** [comment]
Wait for the CI and I will merge :)

**Comment by fzyzcjy** [comment]
@all-contributors please add @alanlzhang for doc

**Pull request merged**